components/style/values/specified/image.rs

```diff
@@ -163,23 +163,54 @@ impl GradientKind {
-        };
-
-        let position = if input.try(|input| input.expect_ident_matching("at")).is_ok() {
-            needs_comma = true;
-            try!(Position::parse(input))
+        let mut needs_comma = true;
+
+        // Ending shape and position can be in various order. Checks all probabilities.
+        let (shape, position) = if let Ok(position) = input.try(parse_position) {
+            (EndingShape::Ellipse(LengthOrPercentageOrKeyword::Keyword(SizeKeyword::FarthestSide)), position)
+        } else if let Ok((first, second)) = input.try(parse_two_length) {
+            let _ = input.try(|input| input.expect_ident_matching("ellipse"));
+            (EndingShape::Ellipse(LengthOrPercentageOrKeyword::LengthOrPercentage(first, second)),
+             input.try(parse_position).unwrap_or(Position::center()))
+        } else if let Ok(length) = input.try(Length::parse) {
+            let _ = input.try(|input| input.expect_ident_matching("circle"));
+            (EndingShape::Circle(LengthOrKeyword::Length(length)),
+             input.try(parse_position).unwrap_or(Position::center()))
+        } else if let Ok(keyword) = input.try(SizeKeyword::parse) {
+            let shape = if input.try(|input| input.expect_ident_matching("circle")).is_ok() {
+                EndingShape::Circle(LengthOrKeyword::Keyword(keyword))
+            } else {
+                EndingShape::Ellipse(LengthOrPercentageOrKeyword::Keyword(keyword))
+            };
+            (shape, input.try(parse_position).unwrap_or(Position::center()))
         } else {
-            Position {
-                horiz_keyword: Some(Keyword::Center),
-                horiz_position: None,
-                vert_keyword: Some(Keyword::Center),
-                vert_position: None,
+            let position = input.position();
+            if let Ok(ident) = input.expect_ident() {
+                match_ignore_ascii_case! { ident,
+                    "circle" => {
+                        let length = input.try(LengthOrKeyword::parse)
+                                     .unwrap_or(LengthOrKeyword::Keyword(SizeKeyword::FarthestCorner));
+                        (EndingShape::Circle(length), input.try(parse_position).unwrap_or(Position::center()))
+                    },
+                    "ellipse" => {
+                        let length = input.try(LengthOrPercentageOrKeyword::parse)
+                                     .unwrap_or(LengthOrPercentageOrKeyword::Keyword(SizeKeyword::FarthestCorner));
+                         (EndingShape::Ellipse(length), input.try(parse_position).unwrap_or(Position::center()))
+                    },
+                    _ => {
+                        // If there is no indent, it should set to default.
+                        // But also it should restore parser's internal state.
+                        needs_comma = false;
+                        input.reset(position);
```

Comment:
I think the best way to structure this broadly is to try to parse the whole thing as if it were an ellipse, first (also eating the comma), else try to parse as a circle (again eating the comma), else default it.